File: 0474-ones-and-zeroes/0474-ones-and-zeroes.java
class Solution {
    public int findMaxForm(String[] strs, int m, int n) {
        int s = strs.length;
        int[] zeros = new int[s];
        int[] ones = new int[s];

        int answer = 1;
        for (int i = 0; i < s; i++) 
         {
            String str = strs[i];
             int z = 0, o = 0;
            for (char c : str.toCharArray())
            {
                if (c == '1')
                {
                    o++;
                }
                else
                {
                    z++;
                }
            }
             zeros[i] = z;
             ones[i] = o;
         }
        

        int[][][] dp = new int[s + 1][m + 1][n + 1];
        
        for (int i = 0; i < s; i++)
            {
                dp[i][0][0] = 1;
                int x = zeros[i];
                int y = ones[i];
                for (int j = 0; j <= m; j++)
                    {
                        for (int k = 0; k <= n; k++)
                        {
                            if (dp[i][j][k] > 0)
                            {
                                if (j + x <= m && k + y <= n)
                                {                    
                                    dp[i + 1][j + x][k + y] = Math.max(dp[i + 1][j + x][k + y], dp[i][j][k] + 1);
                                    answer = Math.max(answer, dp[i + 1][j + x][k + y]);
                                }
                                dp[i + 1][j][k] = Math.max(dp[i + 1][j][k], dp[i][j][k]);
                            }
                        }
                    }
            }

        
        return answer - 1;
    }
}

File: 0474-ones-and-zeroes/0474-ones-and-zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int sr = strs.size();
        vector<vector<int>> counts (sr + 1, vector<int> {0, 0});
        vector<vector<vector<int>>> dp(sr + 1, vector<vector<int>>(m + 1, vector<int>(n + 1, -1)));
        for (int i = 0; i < sr; i++)
        {
            string str = strs[i];
            for (char c : str)
            {
                if (c == '1')
                {
                    counts[i][1]++;
                }
                else
                {
                    counts[i][0]++;
                }
            }
        }

        dp[0][0][0] = 0;
        for (int i = 0; i < sr; i++)
        {
            for (int j = 0; j <= m; j++)
            {
                for (int k = 0; k <= n; k++)
                {
                    if (dp[i][j][k] != -1)
                    {
                        if (j + counts[i][0] <= m && k + counts[i][1] <= n)
                        {
                            dp[i + 1][j + counts[i][0]][k + counts[i][1]] = max(dp[i][j][k] + 1, dp[i + 1][j + counts[i][0]][k + counts[i][1]]);
                        }
                        dp[i + 1][j][k] = max(dp[i][j][k], dp[i + 1][j][k]);
                    }
                }
            }
        }

        int answer = 0;
        for (int i = 0; i <= m; i++)
        {
            for (int j = 0; j <= n; j++)
            {
                answer = max(dp[sr][i][j], answer);
            }
        }
        return answer;
        // 이거 노노야 
    }
};